### PR TITLE
fix: add DynamicPromptsFunction and DynamicActionsFunction to type definitions

### LINKED ISSUE #140: Incorrect types for dynamic action array
# Actual behavior

When [dynamically generating an actions array](https://plopjs.com/documentation/#using-a-dynamic-actions-array) using a `plopfile` written in TypeScript, then TypeScript raises an error:

  ```
 Type '({ componentName }: object) => { type: "add"; path: string; }[]' is not assignable to type 'ActionType<object>[] | CustomActionFunction<object>'.
  Type '({ componentName }: object) => { type: "add"; path: string; }[]' is not assignable to type 'CustomActionFunction<object>'.
    Type '{ type: "add"; path: string; }[]' is not assignable to type 'ActionType<object>[] | Promise<ActionType<object>[]>'.
      Type '{ type: "add"; path: string; }[]' is not assignable to type 'ActionType<object>[]'.
        Type '{ type: "add"; path: string; }' is not assignable to type 'ActionType<object>'.
          Type '{ type: "add"; path: string; }' is missing the following properties from type 'AppendActionConfig<object>': pattern, template, templateFile, force, and 2 more.ts(2322)
index.d.ts(129, 3): The expected type comes from property 'actions' which is declared here on type 'PlopGenerator'
```

```js
import { NodePlopAPI } from 'plop';

export default (plop: NodePlopAPI) => {
  plop.setGenerator('component', {
    description: 'React component',
    prompts: [
      {
        type: 'input',
        name: 'componentName',
        message: "What's the name of your component?"
      },
      // ...
    ],
    actions: ({ componentName }) => [
      {
        type: 'add',
        path: `./${componentName}.js`,
        // ...
      }
    ]
  });
}
```

# Expected behavior

Dynamic action arrays should be typed correctly.

# Context

Plop: 2.5.3
TypeScript: 3.7.3

### LINKED ISSUE #142: action.data function arguments
It seems that when `action.data` is a function, it receives no arguments.

https://github.com/plopjs/node-plop/blob/decdd3f32ecdac65069feeafb07ee8e3ba37be79/src/generator-runner.js#L114-L116

I found the type definition ~~confusing~~ misleading:

https://github.com/plopjs/node-plop/blob/decdd3f32ecdac65069feeafb07ee8e3ba37be79/index.d.ts#L171-L176

Should the data function receive the existing data from the prompts? This could be useful for file loading, etc...

I can make a pull request if this is welcome!

### LINKED ISSUE #162: No typing for functional prompts in generators
Hello, I've been using the library and found a slight issue with the typing for generators. 

There appears to be support for the prompt value of a generator config to be a function: 		
https://github.com/plopjs/node-plop/blob/74f4e08dc44e81945ec4a4f74847bb2319c26436/src/generator-runner.js#L19

But this is not reflected in the typing:
```
  prompts: inquirer.Question[];
```
Let me know if you'd like me to take a crack at this! Should be simple and very similar to the actions typing in that same interface:
```
  actions: Actions | DynamicActionFunction;
``` 

Thanks!

### PR DESCRIPTION
**Pull request by armand1m**
When the PR #37 got merged, Dynamic Prompt functions got implemented and supported, but types ain't got updated.

This also adds a `PromptAnswers` type, which is just an alias to `Record<string, any>`

This PR fixes #162, fixes #142 and fixes #140

### CODE REVIEW

**Comment by armand1m** [comment]
it will need some changes, it looks like the return should not be an array and there is actually an inquirer instance that gets passed as an argument to the prompts call

**Comment by crutchcorn** [comment]
I'm working on a PR to introduce TS typing tests, which may impact this PR. Just a heads up

**Comment by armand1m** [comment]
@crutchcorn noice, thanks for the heads up! I will update your tests once I get this done

**Comment by crutchcorn** [comment]
Alright, so I've moved the `index.d.ts` file to `types` and added a `tests.ts` file. Can you update this PR with the new file location but also write some typing tests as well?

**Comment by armand1m** [comment]
@crutchcorn sure thing!

**Comment by armand1m** [comment]
Fixed merge conflicts, updated the new types definition and bumped @types/inquirer to 6.5.0 to be able to incorporate #167 and avoid another merge conflict. Still need to add the tests to the type definitions.

I also updated `inquirer` to 7.1.0 as a convenience, tests still pass so I believe it is safe. @crutchcorn if you don't feel like bumping this, please let me know so I can revert 🙂

File: types/index.d.ts [hunk 1]
```diff
@@ -58,10 +58,27 @@ interface PlopActionHooks {
...
+export type Prompts = DynamicPromptsFunction | inquirer.Question[]
+	| inquirer.CheckboxQuestion[]
+	| inquirer.ListQuestion[]
+	| inquirer.ExpandQuestion[]
+	| inquirer.ConfirmQuestion[]
+	| inquirer.EditorQuestion[]
+	| inquirer.RawListQuestion[]
+	| inquirer.PasswordQuestion[]
+	| inquirer.NumberQuestion[]
+	| inquirer.InputQuestion[];
```
Comment (armand1m) [inline types/index.d.ts:74]: one thought that comes to my mind here is that this limits the prompts array to always be of a specific type of question.

I'm going to change this to something like this so you can have different types of questions within the same array 

```ts
type PromptQuestion = inquirer.Question
  | /** other types */

type Prompts = DynamicPromptsFunction | PromptQuestion[];
```

Comment (armand1m) [reply types/index.d.ts]: updated this :)

**Comment by armand1m** [comment]
@crutchcorn added some tests and also had to fix some types because they were not matching the current implementation. let me know what you think

**Comment by crutchcorn** [comment]
This looks perfect! Thank you so much!

**Pull request merged**